python/ray/serve/_private/deployment_scheduler.py

```diff
@@ -462,5 +585,95 @@ def key(node_and_num_running_replicas_of_all_deployments):
 
         return replicas_to_stop
 
+    def _schedule_replica_compactly(
+        self,
+        scheduling_request: ReplicaSchedulingRequest,
+    ):
+        replica_name = scheduling_request.replica_name
+        deployment_id = scheduling_request.deployment_id
+        if (
+            scheduling_request.placement_group_bundles
+            and scheduling_request.placement_group_strategy == "STRICT_PACK"
+        ):
+            bundle_resources = [
+                Resources(bundle)
+                for bundle in scheduling_request.placement_group_bundles
+            ]
+            required_resources = sum(bundle_resources, Resources())
+        else:
+            required_resources = Resources(scheduling_request.actor_resources)
+
+        # Get node
+        target_node = self._find_best_available_node(
+            required_resources, self._get_available_resources_per_node()
+        )
```

Comment:
no need to nest this call with `_schedule_replica_compactly`, I think instead the node can be chosen and passed in at the callsite.

btw, isn't the rest of this method call shared completely with the existing scheduling strategy? so it seems to me that we could abstract out just the logic to map replicas to nodes and the rest should not need to diverge across the strategies

Reply:
Good call! I've refactored the code so the scheduling code is shared in `_schedule_replica`, and the information that diverges is passed in at the callsite. PTAL!